crypto/Pos/Crypto/Hashing.hs

```diff
@@ -143,6 +145,15 @@ unsafeAbstractHash
     => a -> AbstractHash algo b
 unsafeAbstractHash = AbstractHash . Hash.hash . Bi.serialize'
 
+-- | Make an AbstractHash from a lazy ByteString. You can choose the phantom
+-- type, hence the "unsafe".
+-- The other 'unsafeAbstractHash' actually serializes the thing, fulfilling
+-- the purpose of the second type parameter, and so is not actually "unsafe".
```

Comment:
`unsafeAbstractHash` is unsafe, because you can hash a `Tx` and get `Hash UpdateProposal`.

Reply:
Oh I see, the second type parameter is `b`, not the `a` with a `Bi a` instance.

Hm, seems weird. Why would we want that? Why would we want to serialize a thing of type `a`, hash it, and then say the hash is of type `b`?  Also, later in this file, `castHash` is defined

```
cast :: forall algo a b . AbstractHash algo a -> AbstractHash algo b
cast = (castHash :: AbstractHash algo Raw -> AbstractHash algo b) . castHash
```

What safety?

Reply:
> Why would we want that? Why would we want to serialize a thing of type a, hash it, and then say the hash is of type b?

Well, one example is your `pureLogic`. You need to create some `StakeholderId` (which is `AddressHash PublicKey`), but you don't want to create a `PublicKey`. This is fine, because `pureLogic` doesn't do anything meaningful. There are other similar cases.
Of course this function shouldn't be used often, its name indicates that.

> Also, later in this file, `castHash` is defined

Actually it's used only in one place and I think it's better to remove this `CastHash` completely.

> What safety?
>  Why bother at all?

@neongreen explained why `AbstractHash` has a phantom parameter: https://github.com/input-output-hk/cardano-sl/pull/2154#discussion_r164922144

Reply:
> Well, one example is your pureLogic. You need to create some StakeholderId (which is AddressHash PublicKey), but you don't want to create a PublicKey

Here I think my new variant is more appropriate: just take the bytestring directly, rather than using the `Bi` instance of something else.

> @neongreen explained why AbstractHash has a phantom parameter

I just mean, if we can define `cast` as I gave above, then why bother with the parameter given that it's so easily bypassed? 

> Actually it's used only in one place and I think it's better to remove this CastHash completely.

Agreed. Either remove casthash, or remove the phantom parameter. Anyway, we're digressing.